src/java/org/apache/cassandra/service/StorageService.java

```diff
@@ -3231,8 +3231,14 @@ static EndpointsByReplica getChangedReplicasForLeaving(String keyspaceName, Inet
 
     public void onJoin(InetAddressAndPort endpoint, EndpointState epState)
     {
-        for (Map.Entry<ApplicationState, VersionedValue> entry : epState.states())
+        Set<Map.Entry<ApplicationState, VersionedValue>> states = epState.states();
+        boolean hasStatusAndPort = states.stream()
+                                         .anyMatch(entry -> entry.getKey() == ApplicationState.STATUS_WITH_PORT);
+        for (Map.Entry<ApplicationState, VersionedValue> entry : states)
         {
+            if (hasStatusAndPort && entry.getKey() == ApplicationState.STATUS) // avoid double-processing this endpoint for things like move/remove
+                continue;
```

Comment:
this is similar to how `Gossiper` handles incoming state changes…
https://github.com/aholmberg/cassandra/blob/trunk/src/java/org/apache/cassandra/gms/Gossiper.java#L1434-L1449

Since `StorageService.onJoin(..)` is getting called from `Gossiper`, should `Gossiper` be handling this?

Reply:
if something equally as simple cannot be applied in Gossiper, then this solution WFM.

Reply:
i was thinking of something like this https://github.com/apache/cassandra/compare/trunk...thelastpickle:mck/trunk_16381

Reply:
I haven't analyzed closely, but I agree that does seem like a better approach. I'm not going to get back to this until Monday so I'm happy to pass to you, or I'll address it then.